core/lib/external_price_api/src/forced_price_client.rs

```diff
@@ -36,35 +43,47 @@ impl ForcedPriceClient {
                 denominator: NonZeroU64::new(denominator).unwrap(),
                 ratio_timestamp: chrono::Utc::now(),
             },
+            previous_numerator: RwLock::new(NonZeroU64::new(numerator).unwrap()),
             fluctuation,
         }
     }
 }
 
 #[async_trait]
 impl PriceAPIClient for ForcedPriceClient {
-    // Returns a ratio which is 10% higher or lower than the configured forced ratio.
+    /// Returns a ratio which is 10% higher or lower than the configured forced ratio,
+    /// but not different more than 3% than the last value
     async fn fetch_ratio(&self, _token_address: Address) -> anyhow::Result<BaseTokenAPIRatio> {
-        if let Some(x) = self.fluctuation {
-            if x != 0 {
-                let mut rng = rand::thread_rng();
-
-                let mut adjust_range = |value: NonZeroU64| {
-                    let value_f64 = value.get() as f64;
-                    let min = (value_f64 * (1.0 - x as f64 / 100.0)).round() as u64;
-                    let max = (value_f64 * (1.0 + x as f64 / 100.0)).round() as u64;
-                    rng.gen_range(min..=max)
-                };
-                let new_numerator = adjust_range(self.ratio.numerator);
-                let new_denominator = adjust_range(self.ratio.denominator);
+        if let Some(fluctation) = self.fluctuation {
+            let mut previous_numerator = self.previous_numerator.write().await;
+            let mut rng = rand::thread_rng();
+            let numerator_range = (
```

Comment:
Adjusting only numerator might not work here. Imagine USDC as a CBT. If one sets up its ratio to 1:2500 - then it'd never be adjusted. Would be great to either adjust both numerator / denominator as it was before - or at least to check which one is adjustable before choosing which one to adjust.

Reply:
I would rather display a warning or silently multiply both denominator and numerator by something like `100`.  Adjusting the denominator by X% has a weird impact on the ratio, because it changes the ratio by factor of `1/(1+X)`. Enforcing next value variation limit becomes harder.

Reply:
Agreed, good point. Can you maybe convert it to float, calculate the next value and then decompose it to numerator / denominator as it's done [here](https://github.com/matter-labs/zksync-era/blob/main/core/lib/external_price_api/src/utils.rs#L7) for example

Reply:
I believe this is going to change denominator in unexpected ways. When you convert a fraction to floating point you loose information and precision.

Reply:
But we would have to do such conversion anyway for any "real" quote providers such as coingecko, coinmarketcap and etc (i.e. they won't be supplying numerator / denominator). Also how much precision is critical for us in the forced client (given that we are allowing it to fluctuate on a fairly random number)

Reply:
This is not something thatI'd block the PR on - but please consider converting the quote to float or multiplying onto some number as you suggested before merging.

Reply:
@ischasny I did add multiplying by 100 if numerator is under 100. I would rebut your argument by that setting ratio to 1:2500 is also something unrealistic as such a setting could not result from converting a "real" quote. Maybe we should go towards changing the forced_price_client config to accept a float for the price itself?

Reply:
Though I understand your point. I think it would be very unexpected for somebody to see the forced price client to set the price to eg. 4/10000 when they set it 1/2500. They could even see something like 7378697629483821/18446744073709551616 depending on the implementation of `Fraction::from(ratio_f64)` that is not actually documented. (The last numbers where taken from a very real python standard implementation)